Baselines and growth

themightychris opened the pull request.

nbey committed:
Add BaselineRating field to StudentCompetency model

nbey committed:
Add baseline input to bulk enroll tool

nbey committed:
Set baseline rating on new level during auto-graduation

- Save the average rating of the concluding StudentCompetency as the baseline rating for the new one being created

nbey committed:
Set baseline when competency has demonstrations for every skill

- Set baseline, if not set yet, when competency has demonstrations for every skill

nbey committed:
Refactor `getCompletionForStudent` method

- Use `StudentCompetency::getCurrentForStudent` method
- Remove `Competency::getCurrentLevelForStudent` & `Competency::getStudentCompetency` methods
- Fix `baselineRating` and add to all responses

nbey committed:
WIP: Refactor getCompletionForStudent to aggregate in PHP

DONE:
- Break out MySQL query into `getDemonstrationData`, `getEffectiveDemonstrationData`, `getDemonstrationsLogged`, `getDemonstrationsComplete`, `getDemonstrationsAverage` methods

TODO:
- Add dynamicFields for each
- Find all the places calling the origin getCompletionForStudent, and replace with newly created methods

nbey committed:
WIP: Refactor getCompletionForStudent to aggregate in PHP

DONE:
- Add dynamicFields for each
- Find all the places calling the origin getCompletionForStudent, and replace with newly created methods

TODO:
- Verify affected CBL exports are accurate

nbey committed:
Add `getTotalSkills` method and private variable

nbey committed:
Implement `getGrowth` method in `StudentCompetency` model

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -136,6 +136,16 @@ public function getSkillIds($forceRefresh = false)
+    private $totalSkills;
+    public function getTotalSkills($forceRefresh = false)
+    {
+        if (!empty($this->totalSkills)) {
```
the difference between `if (!empty($this->totalSkills))` and `if ($this->totalSkills)` is that !empty additionally checks if the variable is defined, and it's impossible for this not to be defined since it's a declared member, so using `empty` here is superfluous

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -136,6 +136,16 @@ public function getSkillIds($forceRefresh = false)
+        }
+
+        return $this->totalSkills = count($this->getSkillIds());
+    }
```
@nbey better pattern is to frame these functions like:

```php
    public function getTotalSkills($forceRefresh = false)
    {
        if ($this->totalSkills === null) {
            $this->totalSkills = count($this->getSkillIds());
        }

        return $this->totalSkills;
    }
```

(`=== null` is needed because `0` is a legit value that would pass `== null`)

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -247,17 +263,18 @@ public function getCompletionForStudent(Student $Student, $level = null, $demons
                 return [
-                    'currentLevel' => $currentLevel,
+                    'currentLevel' => $StudentCompetency->Level,
+                    'baselineRating' => $StudentCompetency->BaselineRating ? floatval($StudentCompetency->BaselineRating) : null,
```
@nbey activerecord should be in charge returning fields as their correct PHP type. If `decimal` doesn't come back as a float that should be added to activerecord the same way it handles ints

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -271,8 +288,9 @@ public function getCompletionForStudent(Student $Student, $level = null, $demons
+            'baselineRating' => $StudentCompetency && $StudentCompetency->BaselineRating ? floatval($StudentCompetency->BaselineRating) : null,
             'demonstrationsAverage' => null,
-            'currentLevel' => $currentLevel
+            'currentLevel' => null
```
@nbey current level still needs to be returned here if it's available

nbey replied:
@themightychris Aren't we deprecating this method?

themightychris replied:
Oh, yeah. As long as the new method doesn't have this regression... This whole thing can just get deleted

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -271,8 +288,9 @@ public function getCompletionForStudent(Student $Student, $level = null, $demons
         return [
             'demonstrationsLogged' => 0,
             'demonstrationsComplete' => 0,
+            'baselineRating' => $StudentCompetency && $StudentCompetency->BaselineRating ? floatval($StudentCompetency->BaselineRating) : null,
```
@nbey this can just be `$StudentCompetency ? $StudentCompetency->BaselineRating : null` once the ActiveRecord returns the correct type on its own

themightychris commented on php-classes/Slate/CBL/Demonstrations/Demonstration.php:
```diff
@@ -112,10 +112,17 @@ public function getCompetencyCompletions()
-            $completions[] = $completion;
+            $StudentCompetency = StudentCompetency::getCurrentForStudent($this->Student);
+
+            $completions[] = [
```
@nbey let's add a method `getCompletion` to `StudentCompetency` that handles returning this data structure since it's a standard of sorts and we're repeating it in a bunch of places.

themightychris replied:
wire up a dynamic field to it too

themightychris commented on php-classes/Slate/CBL/Demonstrations/StudentDashboardRequestHandler.php:
```diff
@@ -122,10 +122,29 @@ public static function handleCompletionsRequest()
+            $StudentCompetency = StudentCompetency::getCurrentForStudent($Competency, $Student);
+
+            if ($StudentCompetency) {
+                $completions[] = [
```
@nbey use `getCompletion()`

themightychris commented on event-handlers/Slate/CBL/Demonstrations/DemonstrationSkill/afterRecordSave/50_complete-competency.php:
```diff
@@ -20,6 +20,11 @@
     ], true);
+} elseif (empty($completion['baselineRating'])) {
+    if ($StudentCompetency = $Competency->getStudentCompetency($Student)) {
+        $StudentCompetency->calculateStartingRating();
```
`calculateStartingRating` is supposed to return a value, not mutate state itself. The writing and saving if it comes back non-null should happen here. The verb "calculate" doesn't sound like a command to mutate state

themightychris commented on event-handlers/Slate/CBL/Demonstrations/DemonstrationSkill/afterRecordSave/50_complete-competency.php:
```diff
@@ -20,6 +20,11 @@
+        'BaselineRating' => $completion['demonstrationsAverage']
     ], true);
+} elseif (empty($completion['baselineRating'])) {
+    if ($StudentCompetency = $Competency->getStudentCompetency($Student)) {
```
`getStudentCompetency` doesn't appear to exist, also this can be consolidated into the if above:

```php
} elseif (empty($completion['baselineRating']) && ($StudentCompetency = StudentCompetency::getCurrentForStudent($Competency, $Student))) {
```

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                $this->Competency->ID
+            ]
+        );
+        $validSkillIds = array_column($validSkills, 'ID');
```
@nbey let's just foreach through `$this->Competency->Skills` and have that build up `$validSkillsId`, we're not gaining any efficiency doing all that json stuff in MySQL

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        );
+
+        // check if each skill has been rated atleast once for the current level
+        if ((int)$ratedSkills !== count($validSkills)) {
```
This check isn't sufficient, there could be ratings for skills with 0 demonstrations required. Use `DB::valuesTable` to get an array of totals indexed by skill ID, and then foreach through `$validSkillIds` and return null when you encounter one where `empty($skillRatings[$skillId])`

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    		'  JOIN `%2$s` d '.
+    		'    ON d.ID = ds.DemonstrationID '.
+    		' INNER JOIN ( '.
+                    'SELECT MIN(demoskill.Created) as Created, SkillID '.
```
Created is when the rating was entered, we want the date of the demonstration.

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            ]
+        );
+
+        $this->BaselineRating = array_sum($demonstrationSkillRatings) / count($validSkills);
```
return the calculation, don't mutate state

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            return $this->demonstrationData = $data;
+        } catch (\Exception $e) {
+            throw $e;
+            return [];
```
The return line will never run, what's the point of catching all exceptions and rethrowing them? Shouldn't this be catching TableNotFoundException and discarding it to return an empty array?

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        $ratedSkills = DB::oneValue(
+
+            'SELECT SUM(IF(skillRatings.total > 0, 1, 0)) as totalRatings '.
+            '  FROM ( '.
```
@nbey there's a mix of SQL indentation styles going on in this new function, you can see the preferred style here: https://github.com/EmergencePlatform/emergence-book/blob/master/recipes/make-a-model-searchable-by-tags.md

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            ' CAST( '.
+                'REPLACE( '.
+            '       IFNULL ( '.
+            '           JSON_EXTRACT(DemonstrationsRequired, CONCAT(\'$."\', %u, \'"\')), '.
```
this CONCAT doesn't need to be done in SQL

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    public function calculateStartingRating($autoSave = true)
     {
-        $completion = $Competency->getCompletionForStudent($Student);
+        $validSkills = DB::allRecords(
```
@nbey read but ignore most of my other feedback in this function, instead start it off with something like this:

```php
$ratings = DB::valuesTable(
    'SkillID',
    'Rating',
    '
    SELECT S.ID AS SkillID,
        IFNULL(
            JSON_EXTRACT(DemonstrationsRequired, \'$."%4$u"\'),
            JSON_EXTRACT(DemonstrationsRequired, "$.default")
        ) AS DemonstrationsRequirements,
        (
            SELECT DS.DemonstratedLevel
              FROM `%3$s` DS
              JOIN `%2$s` D
                ON D.ID = DS.DemonstrationID
             WHERE DS.SkillID = S.ID
               AND DS.TargetLevel = %4$u
               AND DS.DemonstratedLevel > 0
               AND D.StudentID = %5$u
             ORDER BY D.Demonstrated, D.ID
             LIMIT 1
        ) AS Rating
    FROM `%1$s` S
    WHERE S.CompetencyID = %6$u
    HAVING DemonstrationsRequirements > 0
    ',
    [
        Skill::$tableName, // %1$s
        Demonstration::$tableName, // %2$s
        DemonstrationSkill::$tableName, // %3$s
        'Level' => 9, // %4$u
        'StudentID' => 23, // %5$u
        'CompetencyID' => 1 // %6$u
    ]
)
```

This requires CBL-319 be done to eliminate the string->int conversion

themightychris replied:
This will return null values where there is no first rating, so after this you just want to do:

```php
$ratingsCount = count(array_filter($ratings));
if ($ratingsCount != count($ratings) {
    return null;
}

return array_sum($ratings) / $ratingsCount;;
```

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    private $demonstrationData;
+    public function getDemonstrationData()
+    {
+        if (!empty($this->demonstrationData)) {
```
Use `!== null` because an empty array will keep failing `!empty` despite it representing a possible result that shouldn't keep getting refetched

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate',
```
order by DemonstrationID after DemonstartionDate so that multiple demonstrations on the same date get sorted consistently by order entered

nbey replied:
@themightychris `DemonstrationDate` is actually a timestamp column, so there shouldn't be an issue with same day demonstrations.

themightychris replied:
@nbey the column is a timestamp, but the current UI is only a date selector so the time components are mostly the same and we should secondarily sort by ID to ensure consistent sequencing. In the future we might see the time component getting specific values more frequently, especially when we start filling in demonstrations automatically from external systems or primary work completion

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    private $effectiveDemonstrationsData;
+    public function getEffectiveDemonstrationsData()
+    {
+        if (!empty($this->effectiveDemonstrationsData)) {
```
`!== null`

themightychris commented on php-classes/Slate/CBL/Competency.php:
```diff
@@ -293,18 +311,4 @@ public static function getAllByListIdentifier($identifier)
-        return $level ? intval($level) : null;
-    }
-}
+}
```
`getCompletionForStudent` should also be getting deleted from this class

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                $sort[$key]  = $row['DemonstratedLevel'];
+            }
+            // sort demonstrations by highest level
+            array_multisort($sort, SORT_DESC, $demonstrationData);
```
I'm having trouble following what's going on here with `$sort` and `array_multisort`

You just need
```php
protected static function sortEffectiveDemonstrations($a, $b) {
    if ($a['DemonstratedLevel'] == $b['DemonstratedLevel']) {
        return 0;
    }
    return ($a['DemonstratedLevel'] < $b['DemonstratedLevel']) ? -1 : 1;
}
// ...
    foreach ($demonstrationsData as $skillId => &$demonstrationData) {
        uasort($demonstrationData, [__CLASS__,  'sortEffectiveDemonstrations']);
    }
```

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    public function getDemonstrationsLogged()
+    {
+
+        if (!empty($this->demonstrationsLogged)) {
```
`0` is a valid value, use `!== null` and get rid of blank line

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            $skillCount = 0;
+            foreach ($demonstrationData as $demonstration) {
+                if (empty($demonstration['Override'])) {
+                    if (!empty($demonstration['DemonstratedLevel'])) {
```
combine these two ifs

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                    }
+                }
+            }
+            // ?
```
?

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                }
+            }
+            // ?
+            $count += min($skillCount, $Skill->getDemonstrationsRequiredByLevel($this->Level));
```
effective demonstrations is already capped by required

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        }
+
+        if ($this->getDemonstrationsLogged()) {
+            $average = $totalScore / $this->getDemonstrationsLogged();
```
just set `$this->demonstrationsAverage`

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        }
+
+        $effectiveDemonstrationsData = $this->getEffectiveDemonstrationsData();
+        $count = 0;
```
just use `$this->demonstrationsComplete`, initialize it here and return it, we don't need to worry about simultaneous access or anything like that

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        'effectiveDemonstrationsData' => [
+            'getter' => 'getEffectiveDemonstrationsData'
+        ]
+
```
blank line

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        $totalSkills = $this->Competency->getTotalSkills();
+
+        $growthData = array_map(function($demonstrations, $skillId) {
+            if (count($demonstrations) === 1 && empty($this->BaselineRating)) {
```
no need to use `empty` on a model field, it's not at risk of being undefined

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+
+            $lastRating = end($demonstrations);
+            if ($this->BaselineRating) {
+                return $lastRating['DemonstratedLevel'] - floatval($this->BaselineRating);
```
`floatval` won't be needed here after the model field is fixed to return correct type

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                return $lastRating['DemonstratedLevel'] - $firstRating['DemonstratedLevel'];
+            }
+
+        }, $demonstrationData, array_keys($demonstrationData));
```
`$skillId` isn't being used in this map function, so we don't need to pass keys in (which causes the returned array to lose its key=>value association)

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+
+        }, $demonstrationData, array_keys($demonstrationData));
+
+        array_filter($growthData);
```
@nbey array_filter returns a new array it does not modify the passed array. Wrap it around the array_map result above

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+            return null;
+        }
+
+        return array_sum($growthData) / $totalSkills;
```
needs to be memoized

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+        return static::getByWhere(['StudentID' => $Student->ID, 'CompetencyID' => $Competency->ID], ['order' => ['Level' => 'DESC']]);
+    }
+
+    public static function isCurrentLevelComplete($Student, $Competency)
```
needs type hints

nbey committed:
Update `getTotalSkills` method

- Implement 'forceRefresh' logic
- Remove unnecessary empty check
- Enhance function logic pattern

nbey committed:
Remove strict comparison

nbey committed:
Add `getCompletion` method to `StudentCompetency` model

- Add dynamic field `completion`

nbey committed:
Update `competencyCompletions` dynamic field configuration

nbey committed:
Update `calculateStartingRating` method

- Prevent the method from mutating the state of the record

nbey committed:
Update student competency graduation logic

- Fix use statement
- Replace deprecated `getCompletionForStudent` method
- Reimplement `calculateStartingRating` method

nbey committed:
Fix starting rating calculation condition

- Ensure all required skills have at least one rating

nbey committed:
Remove blank line

nbey committed:
Remove MySQL concat function

nbey committed:
Fix MySQL query syntax

nbey committed:
Select first skill rating via `Demonstrated` timestamp column

nbey committed:
Use null comparison instead of checking emptiness

nbey committed:
Ensure demonstration data is consistently ordered

nbey committed:
Handle `TableNotFoundException` errors

nbey committed:
Create method for effective demonstration data sorting

nbey committed:
Combine 'if' statements

nbey committed:
Remove unnecessary code

nbey committed:
Store competency growth calculation

nbey committed:
Add class type hints

nbey committed:
Remove blank line

nbey committed:
Remove deprecated method `getCompletionForStudent`

themightychris commented on php-classes/Slate/CBL/Demonstrations/StudentDashboardRequestHandler.php:
```diff
@@ -122,10 +122,21 @@ public static function handleCompletionsRequest()
+            if ($StudentCompetency) {
+                $completions[] = $StudentCompetency->getCompletion();
+            } else {
+                $completions[] = [
```
@nbey let's create a method `StudentCompetency::getBlankCompletion($Student, $Competency)` for returning this data structure and use it in the teacher request handler too so that as we add to / evolve this data structure there are only two places it is defined and they're both in the same class

themightychris commented on php-classes/Slate/CBL/Demonstrations/StudentDashboardRequestHandler.php:
```diff
@@ -122,10 +122,21 @@ public static function handleCompletionsRequest()
-                'StudentID' => $Student->ID,
-                'CompetencyID' => $Competency->ID
-            ], $Competency->getCompletionForStudent($Student));
+            $StudentCompetency = StudentCompetency::getCurrentForStudent($Competency, $Student);
```
@nbey let's flip the parameter order for getCurrentForStudent so that it matches the name of the class and the order of the index

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+
+    public function calculateStartingRating()
+    {
+        $validSkills = DB::allRecords(
```
@nbey let's use `DB::table('ID', '...', [...])` and then `$validSkillIds = array_keys($validSkills)`

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+                join(', ', $validSkillIds),
+                $this->StudentID
+            ]
+        );
```
@nbey was there any issue with the code I developed here: https://github.com/SlateFoundation/slate-cbl/pull/383#discussion_r132990589

that got us everything we needed in one query

nbey replied:
@themightychris No, for some reason I didn't see that.

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+
+            return $this->demonstrationData = $data;
+        } catch (TableNotFoundException $e) {
+            return [];
```
@nbey this code will result in the empty result of a table being missing not being cached, defeating the purpose of our strict null check to allow for empty arrays as legit values.

Use this pattern instead for all of these functions:

```php
if ($this->demonstrationData === null) {
    try {
        $this->demonstrationData = DB::arrayTable(
            // ...
        );
    } catch (TableNotFoundException $e) {
        $this->demonstrationData = [];
    }
}

return $this->demonstrationData;
```

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+        }
+
+        return $this->demonstrationsLogged = $count;
+
```
blank line

themightychris commented on php-classes/Slate/CBL/StudentCompetency.php:
```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+            $count += $skillCount;
+        }
+
+        return $this->demonstrationsLogged = $count;
```
I mentioned this in the last round, there's no need to use this intermediate variables. Let's work on the variable we're trying to set directly as much as possible to reduce the cognitive load of reading the code

```php

if ($this->demonstrationsLogged === null) {
    $this->demonstrationsLogged = 0;
    foreach (...) {
        $this->demonstrationsLogged += $whatever;
    }
}

return $this->demonstrationsLogged;
```

nbey committed:
Remove blank line

nbey committed:
Reorder `getCurrentForStudent` method arguments

nbey committed:
Add `getBlankCompletion` method

nbey committed:
Index skills array by ID

nbey committed:
Consolidate starting rating calculation query statements

nbey committed:
Ensure skills' `DemonstrationsRequired` column contains integer values

- Convert values during competency/skill import
- Migrate string values to integers

nbey committed:
Add StudentCompetency to use statement

- Reorder use statements

nbey committed:
Return blank completion data if student is not enrolled in competency

nbey committed:
Remove unnecessary `ActiveRecord` field conversion

nbey committed:
Bump `emergence-apikit` submodule version
